File: p67/services/controld/src/lib/sdk.ts
/**
 * @preserve
 * @p67/agent-sdk
 *
 * Agent SDK for P67 platform
 * Provides utilities for workflows to interact with the real world (namely, Snowflake and Cortex services)
 */

import { z } from 'zod';

export const version = '0.1.0';

export const P67ConfigValueSchema = z.object({
    account: z.string().optional(),
    username: z.string().optional(),
    authenticator: z.string().optional(),
    accessUrl: z.string().optional(),
    token: z.string().optional(),
    password: z.string().optional(),
    warehouse: z.string().optional(),
    database: z.string().optional(),
    schema: z.string().optional(),
    email_integration: z.string().optional(),
});

export type P67ConfigValue = z.infer<typeof P67ConfigValueSchema>;

/**
 * P67 configuration schema
 *
 * The core configuration mechanism of the P67 platform.
 */
export const P67ConfigSchema = z.object({
    snowflakeConfig: z.map(z.string(), P67ConfigValueSchema),
});

export type P67Config = z.infer<typeof P67ConfigSchema>;

/**
 * Query execution result
 */
export interface QueryResult {
    statement: unknown;
    rows: unknown[];
}

/**
 * Cortex Analyst API response
 */
export interface CortexAnalystResponse {
    success: boolean;
    data?: unknown;
    error?: string;
}

/**
 * Cortex Agent API response with detailed debugging information
 */
export interface CortexAgentResponse {
    success: boolean;
    status_code: number;
    data?: unknown;
    error?: string;
    request?: {
        url: string;
        headers: Record<string, string>;
        payload: unknown;
    };
    response?: {
        status_code: number;
        headers: Record<string, string>;
    };
}

/**
 * Agent stream event
 */
export interface AgentStreamEvent {
    eventName: string;
    data: unknown;
}

/**
 * Options for calling Cortex Agent
 */
export interface CortexAgentOptions {
    agentDatabase?: string;
    agentSchema?: string;
    agentName?: string;
    parentMessageId?: string;
    onStream?: (event: AgentStreamEvent) => void;
}

/**
 * Snowflake statement binds
 */
export type Bind = string | number | boolean | null;
export type Binds = Bind[];

/**
 * Snowflake statement
 */
export interface SnowflakeStatement {
    sqlText: string;
    binds?: Binds;
}

export interface EmailOptions {
    email_addresses: [string];
    subject: string;
    body: string;
    content_type?: string;
    integration_name?: string;
}

/**
 * Interface for the P67 Agent SDK
 * Defines the public API for interacting with Snowflake and Cortex services
 */
export interface AgentSDK {
    /**
     * Executes a read-only SELECT query against Snowflake
     *
     * Validates that the query is read-only before execution.
     * Allows SELECT, WITH (CTE), SHOW, and DESCRIBE statements.
     * Rejects DML (INSERT, UPDATE, DELETE) and DDL (CREATE, ALTER, DROP) statements for safety.
     *
     * @param query - SQL SELECT query to execute
     * @param config_name - Optional name of the Snowflake config to use. If not provided and only
     *                      one config exists, that config will be used automatically.
     * @returns Query results containing the statement metadata and result rows
     * @throws Error if the query is not read-only, contains multiple statements, or execution fails
     *
     * @example
     * const result = await sdk.executeQueryReadOnly('SELECT * FROM my_table LIMIT 10');
     * console.log(result.rows);
     *
     * @example
     * // Using a specific config
     * const result = await sdk.executeQueryReadOnly(
     *   'SELECT COUNT(*) FROM orders',
     *   'production_config'
     * );
     */
    executeQueryReadOnly(
        stmt: SnowflakeStatement,
        config_name?: string,
    ): Promise<QueryResult>;

    /**
     * Queries Cortex Analyst with a natural language question
     *
     * Uses the Snowflake Cortex Analyst API to convert natural language questions
     * into SQL queries and execute them against your semantic model.
     *
     * @param question - Natural language question to ask (e.g., "What were sales last month?")
     * @param semanticModel - Path to semantic model file (stage path like `@my_stage/model.yaml`).
     *                        Defaults to `CORTEX_ANALYST_SEMANTIC_MODEL` environment variable if not provided.
     * @param config_name - Optional name of the Snowflake config to use. If not provided and only
     *                      one config exists, that config will be used automatically.
     * @returns Response object with `success` boolean, `data` on success, or `error` message on failure.
     *          Never throws - errors are returned in the response object.
     *
     * @example
     * const response = await sdk.queryCortexAnalyst(
     *   'What were the top 5 products by revenue last month?',
     *   '@my_stage/semantic_model.yaml'
     * );
     *
     * if (response.success) {
     *   console.log(response.data);
     * } else {
     *   console.error(response.error);
     * }
     */
    queryCortexAnalyst(
        question: string,
        semanticModel?: string,
        config_name?: string,
    ): Promise<CortexAnalystResponse>;

    /**
     * Calls a Cortex Agent with a question and streams the response
     *
     * Sends a message to a Snowflake Cortex Agent and processes the streaming response.
     * Supports conversational context via `parentMessageId` for multi-turn conversations.
     *
     * @param question - Question or message to send to the agent
     * @param options - Configuration options for the agent call:
     *   - `agentDatabase`: Database containing the agent (defaults to config database)
     *   - `agentSchema`: Schema containing the agent (defaults to config schema)
     *   - `agentName`: Name of the agent (required)
     *   - `parentMessageId`: Parent message ID for conversation continuity (defaults to '0' for new conversations)
     *   - `onStream`: Callback function invoked for each streaming event, receives `{ eventName, data }`
     * @param config_name - Optional name of the Snowflake config to use. If not provided and only
     *                      one config exists, that config will be used automatically.
     * @returns Response object with `success` boolean, `status_code`, `data` on success,
     *          `error` on failure, and `request` debugging info. Never throws - errors are
     *          returned in the response object.
     *
     * @example
     * const response = await sdk.callCortexAgent('What is the weather today?', {
     *   agentDatabase: 'MY_DB',
     *   agentSchema: 'MY_SCHEMA',
     *   agentName: 'weather_agent',
     *   onStream: (event) => {
     *     console.log(`Event: ${event.eventName}`, event.data);
     *   }
     * });
     *
     * if (response.success) {
     *   console.log(response.data.message.content);
     * } else {
     *   console.error(response.error);
     *   console.error('Request:', response.request);
     * }
     *
     * @example
     * // Multi-turn conversation
     * const firstResponse = await sdk.callCortexAgent('Hello', { agentName: 'my_agent' });
     * const secondResponse = await sdk.callCortexAgent('Tell me more', {
     *   agentName: 'my_agent',
     *   parentMessageId: firstResponse.data?.messageId
     * });
     */
    callCortexAgent(
        question: string,
        options?: CortexAgentOptions,
        config_name?: string,
    ): Promise<CortexAgentResponse>;

    /**
     * Sends an email using the Snowflake Email Integration
     *
     * @param email_addresses - Array of email addresses to send the email to
     * @param subject - Subject of the email
     * @param body - Body of the email
     * @param content_type - Content type of the email
     * @param integration_name - Name of the email integration to use. If not provided, the default email integration will be used.
     * @returns Promise that resolves to true if the email is sent successfully, false otherwise
     */
    email(options: EmailOptions, config_name?: string): Promise<boolean>;

    /**
     * Closes the cached Snowflake connection
     *
     * Call this method when shutting down your application to properly cleanup resources
     * and release the connection back to the pool. Safe to call multiple times.
     *
     * @returns Promise that resolves when the connection is closed
     *
     * @example
     * try {
     *   const result = await sdk.executeQueryReadOnly('SELECT * FROM my_table');
     *   // ... process result ...
     * } finally {
     *   await sdk.close();
     * }
     */
    close(): Promise<void>;
}

File: p67/packages/workflow-sdk-python/p67_sdk_stubs/sdk.py
"""
P67 Workflow SDK Interface.

This is a stub file for IDE support only. It contains type signatures
and docstrings but no implementation. The actual implementation is
bundled at build time by 'p67 build'.
"""

from typing import Any, Dict, List, Optional, TypeVar

from p67_sdk.types import (
    QueryResult,
    HttpResponse,
    CortexAnalystResponse,
    CortexAgentResponse,
    CortexCodeOptions,
    CortexCodeResponse,
)

T = TypeVar('T')


class WorkflowSDK:
    """
    P67 Workflow SDK for Python workflows.

    Provides methods to interact with Snowflake, Cortex services,
    and external APIs.

    Example:
        ```python
        def main(sdk: WorkflowSDK) -> dict:
            # Query Snowflake
            result = sdk.execute_query_read_only("SELECT * FROM my_table LIMIT 10")

            # Use Cortex Analyst
            analysis = sdk.query_cortex_analyst(
                "What were total sales last month?",
                semantic_model="@my_db.my_schema.my_stage/model.yaml"
            )

            return {"rows": len(result.rows), "analysis": analysis.data}
        ```
    """

    def __init__(self, config: Dict[str, Any]) -> None:
        """
        Initialize the SDK with configuration from the manifest.

        Note: You don't need to call this directly. The SDK instance
        is created and passed to your main() function automatically.

        Args:
            config: Configuration dict containing snowflakeConfig and parameters
        """
        ...

    def get_parameter(self, name: str, config_name: Optional[str] = None) -> str:
        """
        Get a parameter value from the manifest configuration.

        Parameters are defined in manifest.yaml under the 'parameters' key.

        Args:
            name: The name of the parameter
            config_name: Optional name of the config to use (for multi-config setups)

        Returns:
            The parameter value as a string

        Raises:
            KeyError: If the parameter is not found

        Example:
            ```python
            api_key = sdk.get_parameter("API_KEY")
            threshold = int(sdk.get_parameter("THRESHOLD"))
            ```
        """
        ...

    def get_parameters(self, config_name: Optional[str] = None) -> Dict[str, str]:
        """
        Get all parameters from the manifest configuration.

        Args:
            config_name: Optional name of the config to use

        Returns:
            Dict mapping parameter names to values

        Example:
            ```python
            params = sdk.get_parameters()
            for key, value in params.items():
                print(f"{key}={value}")
            ```
        """
        ...

    def execute_query_read_only(
        self,
        sql_text: str,
        binds: Optional[List[Any]] = None,
        config_name: Optional[str] = None,
    ) -> QueryResult:
        """
        Execute a read-only SQL query against Snowflake.

        Only SELECT, WITH, SHOW, and DESCRIBE statements are allowed.
        DML (INSERT, UPDATE, DELETE) and DDL statements will raise an error.

        Args:
            sql_text: The SQL query to execute
            binds: Optional list of bind parameters for parameterized queries
            config_name: Optional name of the config to use

        Returns:
            QueryResult with:
                - statement: The cursor object
                - rows: List of result rows

        Raises:
            ValueError: If the query is not read-only or contains multiple statements

        Example:
            ```python
            # Simple query
            result = sdk.execute_query_read_only("SELECT * FROM users LIMIT 10")
            for row in result.rows:
                print(row)

            # Parameterized query
            result = sdk.execute_query_read_only(
                "SELECT * FROM orders WHERE customer_id = %s",
                binds=[customer_id]
            )
            ```
        """
        ...

    def query_cortex_analyst(
        self,
        question: str,
        semantic_model: Optional[str] = None,
        config_name: Optional[str] = None,
    ) -> CortexAnalystResponse:
        """
        Query Cortex Analyst with a natural language question.

        Cortex Analyst uses a semantic model to understand your data
        and generate SQL to answer questions.

        Args:
            question: Natural language question to ask
            semantic_model: Path to semantic model file (e.g., "@db.schema.stage/model.yaml")
            config_name: Optional name of the config to use

        Returns:
            CortexAnalystResponse with:
                - success: Whether the query succeeded
                - data: Response data including generated SQL and results
                - error: Error message if failed

        Example:
            ```python
            response = sdk.query_cortex_analyst(
                "What were total sales by region last quarter?",
                semantic_model="@analytics.models.stage/sales_model.yaml"
            )
            if response.success:
                print(response.data)
            else:
                print(f"Error: {response.error}")
            ```
        """
        ...

    def call_cortex_agent(
        self,
        question: str,
        options: Optional[Dict[str, Any]] = None,
        config_name: Optional[str] = None,
    ) -> CortexAgentResponse:
        """
        Call a Cortex Agent with a question.

        Cortex Agents are AI assistants that can answer questions,
        perform tasks, and interact with your data.

        Args:
            question: Question or message to send to the agent
            options: Configuration options:
                - agentDatabase: Database containing the agent
                - agentSchema: Schema containing the agent
                - agentName: Name of the agent
                - parentMessageId: For conversation continuity
            config_name: Optional name of the config to use

        Returns:
            CortexAgentResponse with:
                - success: Whether the call succeeded
                - status_code: HTTP status code
                - data: Response containing the agent's message
                - error: Error message if failed

        Example:
            ```python
            response = sdk.call_cortex_agent(
                "Summarize our Q4 performance",
                options={
                    "agentDatabase": "ANALYTICS",
                    "agentSchema": "AGENTS",
                    "agentName": "SALES_ASSISTANT"
                }
            )
            if response.success:
                print(response.data["message"]["content"])
            ```
        """
        ...

    def email(
        self,
        options: Dict[str, Any],
        config_name: Optional[str] = None,
    ) -> bool:
        """
        Send an email using Snowflake Email Integration.

        Requires a configured email integration in your Snowflake account.

        Args:
            options: Email options:
                - email_addresses: List of recipient addresses
                - subject: Email subject
                - body: Email body content
                - content_type: "text/plain" or "text/html" (default: "text/plain")
                - integration_name: Email integration name (optional if in config)
            config_name: Optional name of the config to use

        Returns:
            True if email was sent successfully, False otherwise

        Example:
            ```python
            success = sdk.email({
                "email_addresses": ["user@example.com"],
                "subject": "Workflow Report",
                "body": "Your daily report is ready.",
                "integration_name": "my_email_integration"
            })
            ```
        """
        ...

    def http_request(self, options: Dict[str, Any]) -> HttpResponse:
        """
        Make an HTTP request to an external service.

        Args:
            options: Request options:
                - url: The URL to request (required)
                - method: HTTP method (default: "GET")
                - headers: Dict of request headers
                - body: Request body (dict/list auto-encoded as JSON)
                - timeout: Timeout in milliseconds (default: 30000)

        Returns:
            HttpResponse with:
                - success: Whether request succeeded
                - status: HTTP status code
                - headers: Response headers
                - data: Response body (parsed as JSON if applicable)
                - error: Error message if failed

        Example:
            ```python
            response = sdk.http_request({
                "url": "https://api.example.com/data",
                "method": "POST",
                "headers": {"Authorization": "Bearer token"},
                "body": {"query": "test"}
            })
            if response.success:
                print(response.data)
            ```
        """
        ...

    def interrupt(
        self,
        payload: Any,
        options: Optional[Dict[str, Any]] = None,
    ) -> T:
        """
        Pause workflow execution and wait for human input.

        The workflow will pause and the payload will be surfaced to
        callers (e.g., a UI). Execution resumes when a response is
        provided via the API.

        Args:
            payload: JSON-serializable value to surface (question, form, etc.)
            options: Optional configuration:
                - timeout: Timeout in milliseconds (None = wait indefinitely)
                - nodeId: Optional node identifier

        Returns:
            The response provided by the human/caller

        Raises:
            TimeoutError: If timeout is reached

        Example:
            ```python
            # Ask for approval
            approved = sdk.interrupt({
                "type": "approval",
                "message": "Proceed with data export?",
                "options": ["yes", "no"]
            })

            if approved == "yes":
                # Continue with export
                pass
            ```
        """
        ...

    def cortex_code(
        self,
        options: CortexCodeOptions,
    ) -> CortexCodeResponse:
        """
        Invoke the Cortex Code CLI to perform agentic coding tasks.

        Runs the ``cortex`` CLI as a subprocess with the given prompt and
        returns the output. Cortex Code is an AI coding agent that can
        read/write files, run commands, search code, and interact with
        Snowflake.

        Args:
            options: CortexCodeOptions containing:
                - prompt: The prompt/instruction to send (required)
                - timeout: Timeout in seconds (default 900 = 15 min)
                - work_dir: Working directory for the process
                - model: Model to use (e.g., 'opus', 'sonnet')
                - allow_all_tool_calls: Skip tool-call confirmations (default False)

        Returns:
            CortexCodeResponse with:
                - success: Whether the invocation succeeded
                - output: Standard output from the cortex process
                - error: Error message if failed
                - exit_code: Process exit code

        Example:
            ```python
            from p67_sdk.types import CortexCodeOptions

            response = sdk.cortex_code(CortexCodeOptions(
                prompt='Read data.csv and summarize the contents.',
                timeout=300,
            ))
            if response.success:
                print(response.output)
            else:
                print(f"Error: {response.error}")
            ```
        """
        ...

    def close(self) -> None:
        """
        Close the Snowflake connection.

        Call this when done to properly release resources.
        Note: This is called automatically when the workflow completes.

        Example:
            ```python
            try:
                result = sdk.execute_query_read_only("SELECT 1")
            finally:
                sdk.close()
            ```
        """
        ...
